common/context/context.cpp

```diff
@@ -53,6 +65,52 @@ kv_db::KeyValueDatabase &MenderContext::GetMenderStoreDB() {
 	return mender_store_;
 }
 
+ExpectedProvidesData MenderContext::LoadProvides() {
+	auto ex_bytes = mender_store_.Read(provides_key_);
+	if (!ex_bytes) {
+		if (ex_bytes.error().code == kv_db::MakeError(kv_db::KeyError, "").code) {
+			// no such data is not an error, just return an empty result
+			ProvidesData ret {};
+			return ExpectedProvidesData(ret);
+		} else {
+			// everything else is an error
+			return ExpectedProvidesData(expected::unexpected(ex_bytes.error()));
+		}
+	}
+
+	auto ex_j = json::Load(common::StringFromByteVector(ex_bytes.value()));
+	if (!ex_j) {
+		return ExpectedProvidesData(expected::unexpected(ex_j.error()));
+	}
+
+	auto ex_children = ex_j.value().GetChildren();
+	if (!ex_children) {
+		return ExpectedProvidesData(expected::unexpected(ex_children.error()));
+	}
+
+	auto children = ex_children.value();
+	if (!all_of(children.cbegin(), children.cend(), [](const json::ChildrenMap::value_type &it) {
+			return it.second.IsString();
+		})) {
+		auto err = json::MakeError(json::TypeError, "Unexpected non-string data in provides");
+		return ExpectedProvidesData(expected::unexpected(err));
+	}
+
+	ProvidesData ret {};
+	for (const auto &it : ex_children.value()) {
```

Comment:
There is a misunderstanding here: The incorrect keys are not stored in the JSON object, they are stored as separate database keys. In other words, in LMDB, you have [three keys](https://github.com/mendersoftware/mender/blob/feature-c%2B%2B-client/common/context.hpp#L46-L57), `artifact-name`, `artifact-group` and `artifact-provides`. The two first keys contain plain values, the last one JSON. But in the end-result, we want to return only one JSON, containing all keys, including the ones which were originally outside in separate keys in the DB.

And then the rename, which you've already done.

For this you will need to use a transaction, because you will need to read three separate keys from the database.

Reply:
Good that I submitted this alone first :)